File: 0567 Permutation in String/0567.cpp
class Solution {
public:
    bool valid(vector<int>& a){
        for (int i = 0; i<a.size(); i++){
            if (a[i] != 0){
                return false;
            }
        }
        return true;
    }
    bool checkInclusion(string s1, string s2) {
        if (s2.size() < s1.size()){
            return false;
        } else {
            vector<int> v(26, 0);

            for (int i = 0; i<s1.size(); i++){
                v[s1[i]-'a']--;
                v[s2[i]-'a']++;
            }

            if (valid(v)){
                return true;
            } else {
                for (int i = s1.size(); i<s2.size(); i++){
                    v[s2[i]-'a']++;
                    v[s2[i-s1.size()]-'a']--;
                    if (valid(v)){
                        return true;
                    }
                }
                return false;
            }
        }
    }
};

File: 0567 Permutation in String/0567.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s2) < len(s1):
            return False
        else:
            v = [0]*26

            for i in range(len(s1)):
                v[ord(s1[i]) - ord('a')] -= 1
                v[ord(s2[i]) - ord('a')] += 1
            
            if set(v) == {0}:
                return True
            else:
                for i in range(len(s1), len(s2)):
                    v[ord(s2[i]) - ord('a')] += 1
                    v[ord(s2[i-len(s1)]) - ord('a')] -= 1
                    
                    if set(v) == {0}:
                        return True

                return False
